packages/astro/src/core/build/plugins/plugin-analyzer.ts

```diff
@@ -317,6 +184,16 @@ export function vitePluginAnalyzer(
 						trackClientOnlyPageDatas(internals, newPageData, clientOnlys);
 					}
 				}
+
+				// When directly rendering scripts, we don't need to group them together when bundling,
+				// each script module is its own entrypoint, so we directly assign each script modules to
+				// `discoveredScripts` here, which will eventually be passed as inputs of the client build.
+				if (options.settings.config.experimental.directRenderScript && astro.scripts.length) {
+					for (let i = 0; i < astro.scripts.length; i++) {
+						const hid = `${id.replace('/@fs', '')}?astro&type=script&index=${i}&lang.ts`;
+						internals.discoveredScripts.add(hid);
+					}
+				}
```

Comment:
I wonder if we could reuse the values in [`hoistedScripts`](https://github.com/withastro/astro/pull/10102/files#diff-2761aafa73765fcf9ead8be801bbe5eb6da3cb305c7b14917299c9fab004e6feR41-R45) by returning it in the `scan` function, looks like they would be identical.